designs/2019-typescript-eslintrc-support/README.md

```diff
@@ -0,0 +1,331 @@
+- Start Date: 2019-11-18
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Gareth Jones ([@g-rath](https://github.com/g-rath))
+
+# First-Class Support For `.eslintrc.ts`
+
+## Summary
+
+Support the ability to consume `.eslintrc.ts` configuration files natively.
+
+## Motivation
+
+Currently configuration settings are validated by JSON Schema, which while
+powerful does not have ubiquitous & high quality support across editors &
+tooling.
+
+It also tends to generate somewhat cryptic messages, especially when using
+complex elements such as combining `oneOf`, `anyOf`, and other such schemas.
+
+Meanwhile, TypeScript is a powerful & popular superset of Javascript that has
+high quality support as a first class language in the majority of editors, and
+is quickly gaining ubiquitous support in general.
+
+By allowing first-class support for `.eslintrc.ts`, the amount of configuration
+problems & general confusion over such experienced by users can be reduced,
+since it would enable us to provide tooling to type configuration files and so
+allow TypeScript to catch ESLint configuration errors at compile time, rather
+than run time.
+
+While we can already hack ways to use `.eslintrc.ts`, none of them lend
+themselves to ease-of-use in the same manner that ESLint encourages, and so
+first-class support for `.eslintrc.ts` will go long way towards encouraging
+community driven typing of ESLint plugins.
+
+Finally, ESLint is being adopted as the linter for TypeScript, meaning
+first-class support for `.eslintrc.ts` will help enable a positive feedback loop
+to help improve the two tools.
+
+There are some potential UX benefits for users that use & understand TypeScript,
+in the form
+
+## Detailed Design
+
+- Adds support for `.eslintrc.ts`
+- Requires uses to install `ts-node` manually adjacent to ESLint in
+  `node_modules`.
+  - Failure to do so with a `.eslintrc.ts` results in ESLint throwing a fatal
+    error
+  - The error message should include understandable plain text instructions on
+    how to fix the error.
+- ESLint does not validate the configuration object for now.
+- `.eslintrc.ts` is second highest priority, after `.eslintrc.js`
+  - This matches the priority of the other supported types, and generally where
+    TypeScript files it for other tools i.e webpack
+
```

Comment:
I'd like to clarify about other pluggable stuff. Does it support `.ts` files in shareable configs, plugins, and custom parsers? For example,

```ts
// .eslintrc.ts
export = {
    extends: [
        "./path/to/a/config.ts",  // (A) file path
        "foo",                    // (B) eslint-config-foo package that may be a `.ts` file
        "plugin:foo/recommended", // (C) `obj.configs.recommended` of esling-plugin-foo
    ],
    parser: "./path/to/a/parser.ts", // (D)
    plugins: ["foo"], // (E)
}
```

From your PoC, it looks to support TS files only in the `(A)` and `(B)`. Is this intentional?

Reply:
Interesting point. I'm all in favor for supporting `plugins` & `parser`, but at the same time would like to get this across the line, so my main question would be that do RFCs require implementation in a single PR? (I assume not, but I could also easily see this broken out into separate RFCs).

Reply:
This is actually an interesting point, that I think maybe should be left for a follow up RFC?

I say that b/c it definitely makes sense to support local `.ts` config files, but ideally you don't want to be *shipping* any ts, i.e plugins, parser, etc b/c then you're require TypeScript via `ts-node` to use them, and thus defeat the whole point.

Reply:
I think for now I'm going to say yes this was intentional, and that I'll add a node to the FAQ:

While open to an RFC for this, it's considered separate to this RFC b/c this isn't about encouraging the shipping of `.ts` based eslint packages; I think the pros and cons to that are different than the ones to this RFC.

Reply:
> would be that do RFCs require implementation in a single PR?

No. There is a precedent, I have implemented #22 with three PRs.

---

I agree that to support local TS files makes sense. But I'm learning toward not supporting TS packages.

Reply:
> I agree that to support local TS files makes sense. But I'm learning toward not supporting TS packages.

Then I think you'll like my new question 😉